artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ClientProducerCreditManagerImpl.java

```diff
@@ -162,26 +162,18 @@ private void addToUnReferencedCache(final SimpleString address, final ClientProd
       unReferencedCredits.put(address, credits);
 
       if (unReferencedCredits.size() > MAX_UNREFERENCED_CREDITS_CACHE_SIZE) {
-         // Remove the oldest entry
-
          Iterator<Map.Entry<SimpleString, ClientProducerCredits>> iter = unReferencedCredits.entrySet().iterator();
-
-         Map.Entry<SimpleString, ClientProducerCredits> oldest = iter.next();
-
-         iter.remove();
-
-         removeEntry(oldest.getKey(), oldest.getValue());
+         while (iter.hasNext()) {
```

Comment:
I would use map.forEach here.


And regarding your concern with always doing this... we could do this through a schedueledExecutor? Schedule a call every X seconds.. and keep it scheduled as long as it's beyond the limit?

Reply:
Regarding using `forEach`...I don't think I can do this because I need to remove an entry from the map I'm inspecting.

I went with the `ScheduledExecutorService`.